### ISSUE TITLE
New bindings

### ISSUE THREAD

**Issue by whitetigle** [fable-compiler/fable-import#7]
Hi there!
So I have new bindings for [waud.js](http://www.waudjs.com) and [paper.js](http://paperjs.org/) libs. My question is simple: where should I put them?

1. in a personal repo?
2. here through a pull request?
3. in a fable child repo?
3. for **paper.js**: in [fable-graphics](https://github.com/fable-compiler/fable-graphics)?

More generally: what's the current policy for adding and registering bindings for Fable? 
Thanks!

**Comment by alfonsogarciacaro** [comment]
We should add this to the README ;) I guess the policy would be:

1. Foremost, **it's the choice of the author**. I guess a personal repo makes more sense when you want to keep more control and publish them yourself and a PR to this repo is better if you want to delegate some responsibilities like publishing or maintenance to fable-compiler.org. Either way is fine, the important thing is to contribute to Fable community 😄 If you choose to keep them in a personal repo, you can send a PR to fable-awesome instead to improve visibility.
2. This repo is mainly intended for pure bindings, that is, without helper code. (There was a explanation for that somewhere but I cannot find it now, maybe @MangelMaxime remembers?) For libraries with helper code, either you can keep them in a repo or commit here only the binding part.

About fable-graphics, we'll probably move the bindings here soon and rename the repo to samples-graphics or similar.

**Comment by MangelMaxime** [comment]
@alfonsogarciacaro Perhpas you are speaking about this issue: #2

**Comment by whitetigle** [comment]
Other questions: 
 

1. when we create new bindings should we add them to nuget as well? 
2. what is the policy regarding tests? What kind of tests should be done?
3. if we put helpers and samples outside, where should we put them? Maybe we could make a single big template à la fable-elmish with sub sections using the imports? For instance, a sub section with pixi and a list of real-life samples? But I think you already may have many ideas in mind 😆

**Comment by MangelMaxime** [comment]
1. I think you should create the nuget too. Then if you want people to be help maintaining the nuget you should add them as Creator on nuget too. Would be nice to have at least 1-2 people able to maintain the nuget to prevent situation where the person stopped using Fable and so the nuget is freeze as no one can update it :)
2. I find testing templates really hard. And tend to say samples is probably one of the easiest way to go. And fix the binding when users encounter problem in it.
3. I am really not sure about separating the Helpers. Otherwise, for splitting the docs/samples in Elmish we have each repo hosting their own docs site. And a central website, linking them all into one place for easier usage.

**Comment by alfonsogarciacaro** [comment]
1. Again, that's up to the author. If you don't want to maintain the Nuget package, just PR it here (together with RELEASE_NOTES.md) and the build script in this repo will create the Nuget package automatically. Later we can make you co-owner of the package if you prefer.
2. I'm not sure about how to test the bindings either, @mike-morr suggested to translate automatically the tests in Definitely Typed. That'd be ideal, though TBH I don't know well how those tests work.
3. My current idea is to move all the bindings **maintained by fable-compiler org** here and convert the other repos to samples like:

```
samples-browser
samples-node
samples-pixi
samples-electron
etc...
```

I think having sample repos around a platform or library is good both for users and maintainers. Users can see several samples about the topic they're interested in, and it's also easier to share dependencies and compile in parallel as in the current `samples` repo.

@MangelMaxime I agree there should not be a single owner of the packages, I'll try to have several owners for each of them just in case :+1:

**Comment by MangelMaxime** [comment]
I am ok with point 1 and 2 of @alfonsogarciacaro anwser :)

For point 2, I am sceptical about the tests. They are better than nothing that's true :) But looking at them they don't test everything methods. Also, looking at them there more or less samples in fact.

After that's true if we take the sample of Pixi they don't need to have a real sample case (like the sample is a platformer). But only test the methods call and that there is no runtime exception.

Perhaps we should indeed encourage people to add tests. And it's possible when you have really simple bindings. Something interacting with a SQL databae we just ask it to connect/disconnect and execute commands nothing more :). But it will not be possible to have test for every bindings.

Another difficulty is for testing database bindings, or really special ones we need to setup an environment etc.

**Comment by alfonsogarciacaro** [comment]
Totally agree, in general we need to improve testing with Fable not only for the bindings but samples, etc. The fable-suave-scaffold has canopy, we can use that or try another more JS-specific tool. It'd be nice to get used to add tests to the samples, because they would be useful as integration tests before releasing new Fable version to make sure there're no regression bugs.

Pinging @Zaid-Ajaj who's been recently playing with qUnit.

**Comment by Zaid-Ajaj** [comment]
@alfonsogarciacaro on elmish side of things: I think QUnit is not adequate enough to do things like integration tests (on it's own, at least), especially not with elmish that is using React. We would need something more specialized like [Jest](https://facebook.github.io/jest/) that can test React components with snapshots and ofcourse make it somehow nice to work with from elmish point of view, @jgrund wrote [Fable.Jest](https://github.com/jgrund/fable-jest) (havn't tested it). [Sinon](sinonjs.org) is also nice for mocks for which @Alxandr wrote the bindings along with bindings for the Ava testing framework [Fable.Ava](https://github.com/YoloDev/Fable.Ava)

QUnit is however nice when doing quick-and-dirty testing and seeing the results very fast and will work just great when you have *pure* functions. for example if you are writing a binding for [Moment.js](https://momentjs.com/) Qunit will be nice and easy to work with

**Comment by jgrund** [comment]
I think a testing story with a modern test framework (something that can run tests in parallel across multiple processes, has module mocking built-in, snapshotting, jsDOM support) would be great, and is probably something devs coming from `react` etc would look for.

Regarding testing things like db bindings, possible to have integration style tests that install dependencies needed, perhaps repos could have their own travis setup?

For unit style tests, we could mock and assert against dependencies calls as well.

**Comment by MangelMaxime** [comment]
Yes that's true. It's we will have one day a lot of Bindings/Test in the repo and creating the setup on travis + running all the tests could take a long time. :) We also need to be careful on that subject too.

**Comment by jgrund** [comment]
@MangelMaxime Indeed. Would be nice to selectively run against changes.

**Comment by MangelMaxime** [comment]
I am closing because we reply to the initial question.

For the tests part, we can open a new issue to discuss it when ready to explore this solution :)
